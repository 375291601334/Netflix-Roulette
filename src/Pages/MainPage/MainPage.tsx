import React, { useEffect, useCallback, ChangeEvent } from 'react';
import { connect } from 'react-redux';
import { actions } from '../../Store';
import { Switch, Route, useLocation, useHistory } from 'react-router-dom';
import { Header, MovieCard, MovieDetails, Navigation, Movie } from '../../Components';
import css from './MainPage.less';

function MainPage(
  {
    loading,
    error,
    movies,
    addMovie,
    updateMovie,
    deleteMovie,
    fetchMovies,
    setFilteringOption,
    setSortingOption,
  }: {
    loading: boolean;
    error: null | Error;
    movies: Movie[];
    addMovie: (movie: Movie) => void;
    updateMovie: (movie: Movie) => void;
    deleteMovie: (movieId: number) => void;
    fetchMovies: (searchString?: string) => void;
    setFilteringOption: (filteringOption: string) => void;
    setSortingOption: (sortingOption: string) => void;
  },
) {
  const location = useLocation();
  const history = useHistory();

  useEffect(
    () => {
      if (location.pathname === '/films' || !movies.length) {
        fetchMovies();
      }
      if (location.pathname === '/search') {
        const queryParam = location.search.match(new RegExp('[?&]searchString=([^&]+).*$'));
        const searchStringParam = queryParam ? queryParam[1] : '';
        fetchMovies(searchStringParam);
      }
    },
    [location],
  );

  const onDeleteMovie = (id: Movie['id']) => {
    deleteMovie(id);
  };

  const changeGenre = useCallback(
    (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
      const container = event.currentTarget;
      const element = event.target as HTMLElement;
      if (element === container) return;

      Array.from(container.getElementsByClassName(css.active)).map(
        element => element.classList.remove(css.active),
      );
      element.classList.add(css.active);

      setFilteringOption(element.dataset.value);
    },
    [],
  );

  const changeSortingOption = useCallback(
    (event: ChangeEvent<HTMLSelectElement>) => {
      setSortingOption(event.currentTarget.value);
    },
    [],
  );

  const getSelectedMovie = (id: string) => {
    const selectedMovie = movies.find(movie => movie.id === + id);
    if (!selectedMovie) {
      history.replace('/films');
    }
    return selectedMovie;
  };

  return (
    <>
      <Switch>
        <Route exact={true} path="/films/:id">
          <MovieDetails getMovie={getSelectedMovie} />
        </Route>
        <Route>
          <Header addMovie={addMovie} />
        </Route>
      </Switch>
      <main>
        <Navigation
          moviesNumber={movies.length}
          changeGenre={changeGenre}
          changeSortingOption={changeSortingOption}
          activeGenreClass={css.active}
        />

          {loading && <div className={css.loader} />}
          {!loading && (
            <>
              {!error && (
                <div className={css.moviesContainer}>
                  {movies.map((movie) => {
                    return (
                      <MovieCard
                        key={movie.id}
                        movie={movie}
                        deleteMovie={onDeleteMovie}
                        editMovie={updateMovie}
                      />
                    );
                  })}
                </div>
              )}
              {!error && !movies.length && (
                <h3 className={css.message}>No movies found</h3>
              )}
              {error && (
                <h3 className={css.message}>Error has occured! Try again in few minutes...</h3>
              )}
            </>
          )}

      </main>
    </>
  );
}

function mapStateToProps({ movies, filtering, sorting }) {
  const { data, loading, error } = movies;
  return {
    loading,
    error,
    movies: data
      .filter(movie => filtering === 'All' || movie.genres.includes(filtering))
      .sort((first, second) => {
        if (first[sorting] > second[sorting]) return -1;
        if (first[sorting] === second[sorting]) return 0;
        if (first[sorting] < second[sorting]) return 1;
      }),
  };
}

const mapDispatchToProps = dispatch => ({
  fetchMovies: (search?: string) => {
    dispatch(actions.fetchMovies());
    return fetch(`http://localhost:4000/movies?limit=18${search && `&searchBy=title&search=${search}`}`)
      .then((response) => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response;
      })
      .then(res => res.json())
      .then((json) => {
        dispatch(actions.fetchMoviesSuccess(json.data));
        return json.data;
      })
      .catch(error => dispatch(actions.fetchMoviesFailure(error)));
  },

  deleteMovie: (movieID) => {
    fetch(`http://localhost:4000/movies/${movieID}`, { method: 'DELETE' })
      .then((response) => {
        if (!response.ok) {
          dispatch(actions.fetchMoviesFailure(response.statusText));
          return;
        }
        dispatch(actions.deleteMovie(movieID));
      });
  },

  updateMovie: (movie) => {
    fetch('http://localhost:4000/movies', { method: 'PUT', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(movie) })
      .then((response) => {
        if (!response.ok) {
          dispatch(actions.fetchMoviesFailure(response.statusText));
          return;
        }
        dispatch(actions.updateMovie(movie));
      });
  },

  addMovie: (movie) => {
    fetch('http://localhost:4000/movies', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(movie) })
      .then((response) => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response;
      })
      .then(res => res.json())
      .then((json) => {
        dispatch(actions.addMovie(json));
        return json.data;
      })
      .catch(error => dispatch(actions.fetchMoviesFailure(error)));
  },

  setFilteringOption: filteringOption => dispatch(actions.setFilteringOption(filteringOption)),
  setSortingOption: sortingOption => dispatch(actions.setSortingOption(sortingOption)),
});

export default connect(mapStateToProps, mapDispatchToProps)(MainPage);
