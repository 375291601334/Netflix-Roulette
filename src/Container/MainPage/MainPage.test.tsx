import React from 'react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import fetchMock from 'jest-fetch-mock';
import configureStore from 'redux-mock-store';
import MainPage, { mapStateToProps } from './MainPage';

const mockStore = configureStore([]);

const movies = [
  {
    id: 123,
    title: 'Movie 1 title',
    tagline: '',
    vote_average: 2.3,
    vote_count: 3.5,
    release_date: '2020-10-13',
    poster_path: 'https://someImage.com',
    overview: 'Movie overview',
    budget: 4365600,
    revenue: 5466453,
    genres: ['Comedy'],
    runtime: 123,
  },
  {
    id: 456,
    title: 'Movie 2 title',
    tagline: '',
    vote_average: 9.3,
    vote_count: 7.5,
    release_date: '2021-10-13',
    poster_path: 'https://someImage.com',
    overview: 'Movie overview',
    budget: 655600,
    revenue: 6453,
    genres: ['Crime'],
    runtime: 846,
  },
];

jest.mock('next/router', () => {
  return {
    useRouter: () => {
      return {
        query: {},
      };
    },
  };
});

describe('MainPage', () => {
  beforeEach(() => {
    fetchMock.doMock();
  });

  test('render MainPage', () => {
    const store = mockStore({
      movies: {
        data: movies,
        loading: false,
        error: null,
      },
      filtering: 'All',
      sorting: 'release_date',
    });

    const { baseElement } = render(
      <Provider store={store}>
        <MainPage />
      </Provider>,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test('render MainPage without movies', () => {
    const store = mockStore({
      movies: {
        data: [],
        loading: false,
        error: null,
      },
      filtering: 'All',
      sorting: 'release_date',
    });

    const { getByText } = render(
      <Provider store={store}>
        <MainPage />
      </Provider>,
    );

    expect(getByText('No movies found')).toBeTruthy();
  });

  test('render MainPage when error occured', () => {
    const store = mockStore({
      movies: {
        data: [],
        loading: false,
        error: new Error('some error'),
      },
      filtering: 'All',
      sorting: 'release_date',
    });

    const { getByText } = render(
      <Provider store={store}>
        <MainPage />
      </Provider>,
    );

    expect(getByText('Error has occured! Try again in few minutes...')).toBeTruthy();
  });

  test('change genre to Comedy', async () => {
    const store = mockStore({
      movies: {
        data: movies,
        loading: false,
        error: null,
      },
      filtering: 'All',
      sorting: 'release_date',
    });

    const { container } = render(
      <Provider store={store}>
        <MainPage />
      </Provider>,
    );

    userEvent.click(container.querySelector('[data-value=Comedy]'));
    expect(store.getActions().pop()).toEqual({ payload: 'Comedy', type: 'SET_FILTERING_OPTION' });
  });

  test('change sorting to revenue', async () => {
    const store = mockStore({
      movies: {
        data: movies,
        loading: false,
        error: null,
      },
      filtering: 'All',
      sorting: 'release_date',
    });

    const { getByRole } = render(
      <Provider store={store}>
        <MainPage />
      </Provider>,
    );

    userEvent.selectOptions(getByRole('navigation').getElementsByTagName('select')[0], 'revenue');
    expect(store.getActions().pop()).toEqual({ payload: 'revenue', type: 'SET_SORTING_OPTION' });
  });

  test('change sorting to revenue', async () => {
    const store = mockStore({
      movies: {
        data: movies,
        loading: false,
        error: null,
      },
      filtering: 'All',
      sorting: 'release_date',
    });

    const { getByRole } = render(
      <Provider store={store}>
        <MainPage />
      </Provider>,
    );

    userEvent.selectOptions(getByRole('navigation').getElementsByTagName('select')[0], 'revenue');
    expect(store.getActions().pop()).toEqual({ payload: 'revenue', type: 'SET_SORTING_OPTION' });
  });

  describe('mapStateToProps', () => {
    test('filter movies by genre', () => {
      const state = {
        movies: {
          data: movies,
          loading: false,
          error: null,
        },
        filtering: 'Comedy',
        sorting: 'release_date',
      };
      expect(mapStateToProps(state)).toEqual({
        loading: state.movies.loading,
        error: state.movies.error,
        movies: movies.filter(movie => movie.genres.includes(state.filtering)),
      });
    });

    test('sort movies by votes if second movie should go first', () => {
      const newMovies = [{ ...movies[0], vote_average: 7 }, { ...movies[1], vote_average: 9 }];
      const state = {
        movies: {
          data: newMovies,
          loading: false,
          error: null,
        },
        filtering: 'All',
        sorting: 'vote_average',
      };
      expect(mapStateToProps(state)).toEqual({
        loading: state.movies.loading,
        error: state.movies.error,
        movies: newMovies.reverse(),
      });
    });

    test('sort movies by votes if first and second movie have equal voices', () => {
      const newMovies = movies.map(movie => movie.vote_average = 9);
      const state = {
        movies: {
          data: newMovies,
          loading: false,
          error: null,
        },
        filtering: 'All',
        sorting: 'vote_average',
      };
      expect(mapStateToProps(state)).toEqual({
        loading: state.movies.loading,
        error: state.movies.error,
        movies: newMovies,
      });
    });

    test('sort movies by votes if first movie should go first', () => {
      const newMovies = [{ ...movies[0], vote_average: 7 }, { ...movies[1], vote_average: 4 }];
      const state = {
        movies: {
          data: newMovies,
          loading: false,
          error: null,
        },
        filtering: 'All',
        sorting: 'vote_average',
      };
      expect(mapStateToProps(state)).toEqual({
        loading: state.movies.loading,
        error: state.movies.error,
        movies: newMovies,
      });
    });
  });
});
