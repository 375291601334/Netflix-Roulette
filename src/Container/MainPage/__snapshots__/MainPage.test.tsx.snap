// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage render MainPage 1`] = `
<body>
  <div>
    <header>
      <button
        class="addButton"
      >
        + ADD MOVIE
      </button>
      <div
        class="searchContainer"
      >
        <label>
          FIND YOUR MOVIE
        </label>
        <div
          class="search"
        >
          <input
            placeholder="What do you want to watch?"
            type="text"
            value=""
          />
          <div
            class="searchButton"
          >
            <a
              href="/search?searchString="
            >
              SEARCH
            </a>
          </div>
        </div>
      </div>
    </header>
    <main>
      <nav>
        <ul>
          <li
            class="active"
            data-value="All"
          >
            All
          </li>
          <li
            data-value="Documentary"
          >
            Documentary
          </li>
          <li
            data-value="Comedy"
          >
            Comedy
          </li>
          <li
            data-value="Horror"
          >
            Horror
          </li>
          <li
            data-value="Crime"
          >
            Crime
          </li>
        </ul>
        <div
          class="sorting"
        >
          <label>
            SORT BY
          </label>
          <select>
            <option
              value="release_date"
            >
              Release date
            </option>
            <option
              value="revenue"
            >
              Revenue
            </option>
            <option
              value="vote_average"
            >
              Votes
            </option>
          </select>
        </div>
      </nav>
      <div
        class="postNavigation"
      >
        <div
          class="counting"
        >
          2
           movies found
        </div>
        <div
          class="mobileSorting"
        >
          <label>
            SORT BY
          </label>
          <select>
            <option
              value="release_date"
            >
              Release date
            </option>
            <option
              value="revenue"
            >
              Revenue
            </option>
            <option
              value="vote_average"
            >
              Votes
            </option>
          </select>
        </div>
      </div>
      <div
        class="moviesContainer"
      >
        <div
          class="cardContainer"
        >
          <span
            class="menuIcon"
          >
            ⋮
          </span>
          <div
            class="menu"
          >
            <button
              class="closeButton"
            >
              <span
                class="closeIcon"
              >
                +
              </span>
            </button>
            <div
              class="option"
            >
              Edit
            </div>
            <div
              class="option"
            >
              Delete
            </div>
          </div>
          <div
            class="card"
          >
            <img
              alt="Movie 2 title poster"
              src="https://someImage.com"
            />
            <div
              class="mainInfoContainer"
            >
              <h4>
                Movie 2 title
              </h4>
              <span
                class="date"
              >
                2021
              </span>
            </div>
            <p>
              Crime
            </p>
          </div>
        </div>
        <div
          class="cardContainer"
        >
          <span
            class="menuIcon"
          >
            ⋮
          </span>
          <div
            class="menu"
          >
            <button
              class="closeButton"
            >
              <span
                class="closeIcon"
              >
                +
              </span>
            </button>
            <div
              class="option"
            >
              Edit
            </div>
            <div
              class="option"
            >
              Delete
            </div>
          </div>
          <div
            class="card"
          >
            <img
              alt="Movie 1 title poster"
              src="https://someImage.com"
            />
            <div
              class="mainInfoContainer"
            >
              <h4>
                Movie 1 title
              </h4>
              <span
                class="date"
              >
                2020
              </span>
            </div>
            <p>
              Comedy
            </p>
          </div>
        </div>
      </div>
    </main>
  </div>
</body>
`;
